File: examples/node_examples/test_mtrace.js
const http = require("http");
var mtrace = require("mtrace");

const hostname = "127.0.0.1";
const port = 3000;

/**
 * This is an example using the wrapped mtrace functions.
 *
 * PREREQUISITES:
 * - compile the node module first (see main README)
 * - install the node module by running "npm install pathToModule" in this directory
 * - set the env variable MALLOC_TRACE, the target log file
 *   like MALLOC_TRACE=~/mtrace.out
 * 
 * RUN: node test_mtrace.js
 * Access the server via browser.
 * After running the program there should be content in the specified log file.
 */
const server = http.createServer((req, res) => {
  console.log(`Environment variable MALLOC_TRACE has value: ${process.env["MALLOC_TRACE"]}`);
  mtrace.mtrace();
  allocate();
  mtrace.muntrace();

  res.statusCode = 200;
  res.setHeader("Content-Type", "text/plain");
  res.end("Done. mtrace results should be present.");
});

function allocate() {
  for (var i = 0; i < 10; i++) {
    const url = new URL("http://example." + i);
    console.log(`${url.href}`);
  }
}

server.listen(port, hostname, () => {
  console.log(`Server running at http://${hostname}:${port}/`);
});

File: examples/native_examples/test_mtrace.cc
#include <stdlib.h>
#include <mcheck.h>

/* This is an c++ example using the native mtrace functions.
 It can be used to verify that mtrace actually works on the current system.

 PREREQUISITE: set the env variable MALLOC_TRACE, the target log file
 like MALLOC_TRACE=~/mtrace.out

 COMPILE: g++ -o ./test test_mtrace.cc
 RUN: ./test
 
 After running the program there should be content in the specified log file.
  */
int main(void)
{

	mtrace(); /* Starts the recording of memory allocations and releases */
	double *a = NULL;
	a = (double *)malloc(sizeof(double));
	if (a == NULL)
	{
		return 1; // return error codeS
	}
	muntrace();
	return 0; 
}